log: use fake stacktrace when test-exiting

tbg opened the pull request:
Having stack traces littered across tests has been a long-standing pet
peeve of mine (and presumably others). We search through logs more often
than we would like to admit, and having to skip a number of intentional
stack traces is annoying.

Make it so that whenever the logging exit func is mocked, the stack
trace indicates that.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/27329)
<!-- Reviewable:end -->

benesch reviewed (approved):
It just occurred to me: I think it would be cleaner to split this into two APIs. What we really want is an atexit callback for non tests, and an exit func override for tests. The atexit callback should not be able to prevent the exit, nor should it be responsible for issuing the `os.Exit` call; it should just run some cleanup code.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> though

> Reviewed 11 of 11 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27329#-:-LH3_dr9-ARPhzsxyffk:b4d1zj3)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/util/log/clog.go, line 695 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27329#-LH3ZGmICU9GMisLm9hC:-LH3ZGmICU9GMisLm9hD:b-5iqcmk) ([raw file](https://github.com/cockroachdb/cockroach/blob/21f2f6ac7c04832690892320cdb7b8b129ce41d6/pkg/util/log/clog.go#L695)):*
> ```Go
> 	exitOverride struct {
> 		f         func(int) // overrides os.Exit when non-nil; testing only
> 		hideStack bool      // hides stack trace; only in effect when f is nil
> ```

when f is non-nil, no?
___
*[pkg/util/log/clog.go, line 857 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27329#-LH3_RIj6nK4sg2bmQzu:-LH3_RIj6nK4sg2bmQzv:b-seqliw) ([raw file](https://github.com/cockroachdb/cockroach/blob/21f2f6ac7c04832690892320cdb7b8b129ce41d6/pkg/util/log/clog.go#L857)):*
> ```Go
> 			// We're running in a test. It's a little dirty that we're detecting
> 			// this here, but we definitely don't want to litter stacktraces in
> 			// the test output when there isn't anything going wrong.
> ```

I'm confused by this comment. f can be non-nil outside of tests.
___
*[pkg/util/log/clog_test.go, line 97 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27329#-LH3_ApC9LraDtgUed7d:-LH3_ApC9LraDtgUed7e:bvz1aya) ([raw file](https://github.com/cockroachdb/cockroach/blob/21f2f6ac7c04832690892320cdb7b8b129ce41d6/pkg/util/log/clog_test.go#L97)):*
> ```Go
> // setFlags resets the logging flags and exit function to what tests expect.
> func setFlags() {
> 	SetExitFunc(false, nil)
> ```

This would read much more nicely as `ResetExitFunc()`.
___
*[pkg/util/log/clog_test.go, line 701 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27329#-LH3_D-2-wf_ICbMGpFk:-LH3_D-2-wf_ICbMGpFl:b-6ua5jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/21f2f6ac7c04832690892320cdb7b8b129ce41d6/pkg/util/log/clog_test.go#L701)):*
> ```Go
> 	setFlags()
> 	logging.stderrThreshold = Severity_NONE
> 	SetExitFunc(false, func(int) {})
> ```

`/* hideStack */`
___
*[pkg/util/log/log.go, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27329#-LH3ZisG71N2zTBPun0R:-LH3ZisG71N2zTBPun0S:b-v0qkta) ([raw file](https://github.com/cockroachdb/cockroach/blob/21f2f6ac7c04832690892320cdb7b8b129ce41d6/pkg/util/log/log.go#L53)):*
> ```Go
> 	if f == nil && hideStack {
> 		panic("can't pass hideStack when resetting logger")
> 	}
> ```

Maybe introduce `ResetExitFunc()` that does the right thing?


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
TFTR! I introduced `ResetExitFunc` but stopped shy of the two separate APIs (started but realized it wasn't worth it for the three obscure users).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27329#-:-LH4buWQ9vq25zrWz98T:b-y62k25)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/util/log/clog.go, line 695 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27329#-LH3ZGmICU9GMisLm9hC:-LH4bVfs2cP_uuF1pOG5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/21f2f6ac7c04832690892320cdb7b8b129ce41d6/pkg/util/log/clog.go#L695)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

when f is non-nil, no?
</blockquote></details>

Done.
___
*[pkg/util/log/clog.go, line 857 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27329#-LH3_RIj6nK4sg2bmQzu:-LH4bXA25WVQalwksxz3:bp9sske) ([raw file](https://github.com/cockroachdb/cockroach/blob/21f2f6ac7c04832690892320cdb7b8b129ce41d6/pkg/util/log/clog.go#L857)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

I'm confused by this comment. f can be non-nil outside of tests.
</blockquote></details>

Done. The comment predates my finding out this was the case.


<!-- Sent from Reviewable.io -->

benesch reviewed (approved):
> Reviewed 7 of 7 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27329)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

tbg committed:
log: use fake stacktrace when test-exiting

Having stack traces littered across tests has been a long-standing pet
peeve of mine (and presumably others). We search through logs more often
than we would like to admit, and having to skip a number of intentional
stack traces is annoying.

Make it so that whenever the logging exit func is mocked, the stack
trace indicates that.

Release note: None

tschottdorf commented:
bors r=benesch
